grape.py

```diff
@@ -67,58 +73,51 @@ def grape(H0, Hk, u_0, rho_0, C, T, alpha=1e-3, epsilon=1e-6, max_iter=1000, min
     return threshold, u_kj, rhoj
 
 
-def cal_Uj(H0, Hk, delta_t, u_kj):
+def cal_Uj(H0, Hk, delta_t, u_kj):    #u_kj mxN, Hk mxnxn, Uj Nxnxn matrix
     m, N = np.shape(u_kj)
     n = H0.shape[0]
 
-    Uj = np.ndarray((N, n, n), dtype=np.complex128)
-    for j in range(N):
-        sigma = np.zeros((n, n), dtype=np.complex128)
-        for k in range(m):
-            sigma += u_kj[k, j] * Hk[k]
-
-        Uj[j] = expm(-1j * delta_t * (H0 + sigma))
-
+    sigma = np.tensordot(u_kj, Hk, axes=([0], [0]))  # Nxnxn
+    Uj = expm(-1j * delta_t * (H0 + sigma))
     return Uj
 
 
 def cal_rhoj(Uj, rho_0):
     N = np.shape(Uj)[0]
     n = np.shape(Uj)[1]
+    rho_0 = np.array(rho_0)
 
     rhoj = np.ndarray((N, n, n), np.complex128)
-    for j in range(N):
-        rho = np.array(rho_0)
-        for i in range(j + 1):
-            rho = Uj[i] @ rho @ (Uj[i].conj().T)
-        rhoj[j] = rho
+    rhoj[0] = Uj[0] @ rho_0 @ (Uj[0].conj().T)
+    for j in range(1,N):
+        rhoj[j] = Uj[j] @ rhoj[j-1] @ (Uj[j].conj().T)

```

Comment:
It's fine to use such method, but I do recommand that you shall use `np.cumprod` to futher utlize the GPU acceleration. Same with `cal_lambdaj`